articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /engineering-education/what-is-an-abi/
+title: What is an ABI?
+description: The application binary interface (ABI) orchestrates how code binaries works together. 
+author: zack-jorquera
+date: 2020-08-27T00:00:00-13:00
+topics: 
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/what-is-an-abi/hero.png
+    alt: hero image ABI application binary interface library int rust c++ cpp ffi FFI
+---
+Example intro here
+<!--more-->
+
+# What is an ABI
+
+An ABI is an application binary interface. It has to do with the implementation details of code i.e. what code turns into when it's compiled. Sometimes ABIs are in place to conform to hardware/kernel requirements but most of the time they are to make sure that two pieces of binary code can work together (like using a pre-compiled library).
+
+We can imagine why this might be important: let's say we are using a library with an API that requires the caller to pass some arguments to a function. Now, how does your code know how to pass those arguments to the library function? Generally, we would use the stack or registers but we want to guarantee that both our code and the library know what to use and in what order. This is exactly the problem that an ABI solves.

```

Comment:
We can imagine why this might be important: let's say we are using a library with an API that requires the caller to pass some arguments to a function. Now, how does your code know how to pass those arguments to the library function? Generally, we would use the stack or registers, but we want to guarantee that both our code and the library know what to use and in what order. This is the problem an ABI solves.

Reply:
What do you want me to change?

Reply:
O I see. you added a comma after _registers_